// gatsby/dist/redux/reducers/index.js
"use strict";

var _nodes = require("./nodes");

var _pages = require("./pages");

var _redirects = require("./redirects");

var _schema = require("./schema");

var _staticQueryComponents = require("./static-query-components");

var _status = require("./status");

var _webpack = require("./webpack");

var _pageData = require("./page-data");

var _themes = require("./themes");

var _webpackCompilationHash = require("./webpack-compilation-hash");

var _reducer = require("gatsby-cli/lib/reporter/redux/reducer");

var _lastAction = require("./last-action");

const nodesByType = require(`./nodes-by-type`);

/**
 * @property exports.nodesTouched Set<string>
 */
module.exports = {
  program: require(`./program`),
  nodes: _nodes.nodeReducer,
  nodesByType: nodesByType,
  resolvedNodesCache: require(`./resolved-nodes`),
  nodesTouched: require(`./nodes-touched`),
  lastAction: _lastAction.lastAction,
  flattenedPlugins: require(`./flattened-plugins`),
  config: require(`./config`),
  schema: _schema.schemaReducer,
  pages: _pages.pagesReducer,
  status: _status.statusReducer,
  componentDataDependencies: require(`./component-data-dependencies`),
  components: require(`./components`),
  staticQueryComponents: _staticQueryComponents.staticQueryComponentsReducer,
  jobs: require(`./jobs`),
  jobsV2: require(`./jobsv2`),
  webpack: _webpack.webpackReducer,
  webpackCompilationHash: _webpackCompilationHash.webpackCompilationHashReducer,
  redirects: _redirects.redirectsReducer,
  babelrc: require(`./babelrc`),
  schemaCustomization: require(`./schema-customization`),
  themes: _themes.themesReducer,
  logs: _reducer.reducer,
  inferenceMetadata: require(`./inference-metadata`),
  pageDataStats: require(`./page-data-stats`),
  pageData: _pageData.pageDataReducer
};
